### PR TITLE
Add option of having device functions shared across opencl kernels

### PR DESCRIPTION
**Pull request by t4c1**
## Summary

Allows selecting which common functionality is available in each opencl kernel. Existing kernel helpers are also refactored, so that they are not in all kernels by default.

## Tests

No new tests. Existing tests suffice.

## Side Effects

There are no more global opencl kernel helpers. Each kernel must specify what common functionality it needs.

## Checklist

- [x] Math issue #1200

- [x] Copyright holder: Tadej Ciglarič, University of Ljubljana

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by SteveBronder** [comment]
Can I get some context on this one? Why is having these the same across each kernel bad?

**Comment by rok-cesnovar** [comment]
We were discussing with @t4c1 how to add helper functions like `digamma` that he is using for the GPU GLM functions [here](https://github.com/bstatcomp/math/blob/gpu_glm2/stan/math/opencl/kernels/neg_binomial_2_log_glm_lpmf.hpp#L19).

The way we have things setup now, the two options would be to

-  copy/paste the entire digamma function to each kernel where its needed
-  add it to the helpers file which would mean that this would be compiled with every kernel, which seems unnecessary

For the amount of helper macros, defines, etc. we have right now, this PR may seem like overkill, but once we have a few more of functions like digamma, this seems like a good solution, I think.

**Comment by SteveBronder** [comment]
Word makes sense!

File: stan/math/opencl/kernel_cl.hpp [hunk 1]
```diff
@@ -146,7 +158,17 @@ struct local_range_kernel {
    */
   local_range_kernel(const char* name, const char* source,
                      const std::map<const char*, int> options = {})
-      : make_functor(name, source, options) {}
+      : make_functor(name, {source}, options) {}
+  /**
+   * Creates kernels that need access to defining the global thread
+   * siez and the thread block size.
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernel_cl.hpp]: size

File: stan/math/opencl/opencl_context.hpp [hunk 2]
```diff
@@ -379,6 +379,7 @@ class opencl_context {
     return {opencl_context_base::getInstance().platform_};
   }
 };
+
```
Comment (rok-cesnovar) [inline stan/math/opencl/opencl_context.hpp]: can you remove this or is it autoformatter trying to work his magic?

File: stan/math/opencl/kernels/helpers.hpp [hunk 3]
```diff
@@ -11,7 +11,7 @@ namespace opencl_kernels {
 /*
  * Defines some helper macros for the kernels
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernels/helpers.hpp]: Please expand these comments with a little infor on what the helpers help at. Same for the thread block helpers. Something like "Helpers for readable matrix element access."

File: stan/math/opencl/kernel_cl.hpp [hunk 4]
```diff
@@ -25,20 +25,22 @@ namespace opencl_kernels {
  * Compile an OpenCL kernel.
  *
  * @param name The name for the kernel
- * @param source A string literal containing the code for the kernel.
+ * @param sources A std::vector of strings containing the code for the kernel.
  * @param options The values of macros to be passed at compile time.
  * @note The macros defined in kernels/helpers.hpp are included in the kernel
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernel_cl.hpp]: This note can probably be removed now?

**Comment by SteveBronder** [comment]
I'll take a look today

File: stan/math/opencl/kernels/helpers.hpp [hunk 5]
```diff
@@ -9,9 +9,9 @@ namespace math {
 namespace opencl_kernels {
 
 /*
- * Defines some helper macros for the kernels
+ * Defines helper macros for common matrix indexing operations
  */
-static const char* helpers =
+static const char* indexing_helpers =
```
Comment (seantalts) [inline stan/math/opencl/kernels/helpers.hpp]: I like that you're leaving them all in the same file. It's nice to think about the use-case of someone new to the code trying to get familiar and write their own kernels or edit existing ones. 👍

File: stan/math/opencl/kernel_cl.hpp [hunk 6]
```diff
@@ -25,20 +25,20 @@ namespace opencl_kernels {
...
  *
  * @param name The name for the kernel
- * @param source A string literal containing the code for the kernel.
+ * @param sources A std::vector of strings containing the code for the kernel.
  * @param options The values of macros to be passed at compile time.
- * @note The macros defined in kernels/helpers.hpp are included in the kernel
- *  compilation for ease of writing and reading kernels.
  */
-inline auto compile_kernel(const char* name, const char* source,
+inline auto compile_kernel(const char* name, std::vector<const char*> sources,
```
Comment (seantalts) [inline stan/math/opencl/kernel_cl.hpp]: Very elegant! Excellent :)

**Comment by seantalts** [comment]
I'm a fan of the way this is done. I am not 100% sure it feels totally motivated yet but I'm willing to go with y'all's intuitions / knowledge of future plans to motivate it, and it's still very clean.

File: stan/math/opencl/kernel_cl.hpp (hunk 6 above)
Comment (SteveBronder) [reply stan/math/opencl/kernel_cl.hpp]: Agree! I think Pretty sure we can make that
 `const std::vector<const char*>& sources`

Comment (SteveBronder) [reply stan/math/opencl/kernel_cl.hpp]: ^ I think that's true most places. @t4c1 while you are here would you mind cleaning up the signatures for stuff we can make const &?

Comment (t4c1) [reply stan/math/opencl/kernel_cl.hpp]: Of course. I'm also making `std::map<const char*, int> options` a reference.

**Pull request merged**